### PR TITLE
Speedup JVP for triangular solve

### PR DESCRIPTION
**Pull request by shoyer**
There is still room for improvement, e.g., by combining the currently separate
JVPs for a and b into a single expression (which will allow for saving an inner
triangular solve when both arguments are being differentiated), but this is
already significantly faster in the typical case of only solving a single
vector.

On my laptop's CPU, I measure 2.98 ms before vs 1.18 ms after on a 500x500 matrix (the forward pass is 527 µs):

    import numpy as onp
    import jax.numpy as np
    import jax
    rs = onp.random.RandomState(0)
    a = rs.randn(500, 500)
    b = rs.randn(500)

    @jax.jit
    def loss(a, b):
      return np.sum(jax.scipy.linalg.solve_triangular(a, b))
    grad = jax.jit(jax.grad(loss))

    %timeit jax.device_get(grad(a, b))

### CODE REVIEW

File: jax/lax_linalg.py [hunk 1]
```diff
@@ -319,18 +320,28 @@ def triangular_solve_jvp_rule_a(
...
+    # Note: triangular_solve is about the same cost as matrix multplication
+    # (~n^2 FLOPs for matrix/vector inputs). Order these operations in
+    # whichever order is cheaper.
+    if m > n:
+      return a_inverse(dot(g_a, ans))  # A^{-1} (∂A X)
+    else:
+      return dot(a_inverse(g_a), ans)  # (A^{-1} ∂A) X
   else:
-    return dot(ans, tmp)
+    return dot(ans, a_inverse(g_a))
```
Comment (hawkinsp) [inline jax/lax_linalg.py]: Doesn't the same apply in the `left_side == False` case?

Comment (shoyer) [reply jax/lax_linalg.py]: Yes, I suspect so, though I was not confident how the math works out to compute `dot(ans, a_inverse)` in a matrix-free way. Given that `left_side=True` is most common, maybe I'll just add a TODO note here for now?

**Review by hawkinsp — APPROVED** [review]

File: jax/lax_linalg.py (hunk 1 above)
Comment (shoyer) [reply jax/lax_linalg.py]: Actually, this turned out to be pretty straightforward once I thought about it. Please take another look.

In particular, I'm a little nervous about test coverage. We don't have any direct test coverage for triangular_solve's JVP (the solve_triangular tests only cover `left_side=True`). There's some indirectly coverage via the JVP rules for cholesky and LU -- maybe that's good enough?

Comment (hawkinsp) [reply jax/lax_linalg.py]: Well, more testing is better.

I think the only reason the test isn't there is that `scipy`'s API doesn't expose this option. (We expose it mostly because LAPACK does, and it's sometimes pretty handy.)

We could test the primitive directly instead of or in addition to the `scipy` wrapper.

Up to you.

Comment (shoyer) [reply jax/lax_linalg.py]: I switched over the gradient tests to use `lax_linalg`

File: tests/linalg_test.py [hunk 2]
```diff
@@ -750,33 +751,40 @@ def testSolveTriangular(self, lower, transpose_a, unit_diagonal, lhs_shape,
...
+       "lower": lower, "transpose_a": transpose_a, "conjugate_a": conjugate_a,
+       "unit_diagonal": unit_diagonal, "left_side": left_side,
+       "a_shape": a_shape, "b_shape": b_shape, "dtype": dtype, "rng": rng}
       for lower in [False, True]
       for unit_diagonal in [False, True]
       for dtype in float_types + complex_types
-      for transpose_a in (
-        [0, 1] if onp.issubdtype(dtype, np.floating) else [0, 1, 2])
-      for lhs_shape, rhs_shape in [
-          ((4, 4), (4,)),
```
Comment (hawkinsp) [inline tests/linalg_test.py]: I note we no longer test the scipy wrapper at all, and the one useful thing it does is support vectors on the RHS.

Perhaps add another test just for that case?

**Review by hawkinsp — APPROVED** [review]

**Review by hawkinsp — APPROVED** [review]
